### ISSUE TITLE
[no-explicit-any] allow usage in extends with invariant generic types?

### ISSUE THREAD

**Issue by P4** [typescript-eslint/typescript-eslint#3757]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I think I might have found a possible edge-case where the rule should to allow using any, because neither replacement (`unknown` nor `never`) seems to work.

I have an instance of a generic class `MatTreeFlattener` from Angular Material that had it's type parameters inferred by the compiler. Now I want to extract the type of one of the type parameters, which can be done using conditional types and `infer` keyword:
```TS
type FlatNode = typeof flattener extends MatTreeFlattener<any, infer F, any> ? F : never;
```
I'm only interested in the second parameter so I use `any` to ignore the others. ESLint obviously doesn't like this and suggests using either `unknown` or `never`. However if I actually follow its advice then the code stops working - the condition no longer matches and the resulting type is `never` instead of the type inferred as `F`. 

This can be worked around by using `infer` on the other parameters (though that triggers `no-unused-vars`), but the same case will occur in `extends` clause on a generic type if I try to make a helper similar to `ReturnType`:
```TS
type Flat<FF extends MatTreeFlattener<any, any, any>> = 
    FF extends MatTreeFlattener<any, infer F, any> ? F : never;
```
This won't compile if I replace `any` with something else, and `infer` cannot be used in extends outside of conditional types.

**Repro**

```JSON
{
    "rules": {
        "@typescript-eslint/no-explicit-any": "error"
    }
}
```

```TS
// Minimal version of MatTreeFlattener from @angular/material/tree, has the same issue as the original
class Flattener<T, F> {
    constructor(
        readonly transform: (node: T, level: number) => F, 
        readonly children: (node: T) => T[]
    ) {}
}

interface Entity {
    id: number; name: string; children: Entity[];
}

const flattener = new Flattener(
  ({ id, name }: Entity, level) => ({ id, name, level }),
  (org: Entity) => org.children
);

type FlatNode = typeof flattener extends Flattener<any, infer F> ? F : never;
//                                                 ^^^ Unexpected any. Specify a different type
```

[TypeScript Playground Link](https://www.typescriptlang.org/play?target=2#code/PTAEFkEsDtIWwIYBtQDcCmAnAzpA9tKHgGYQIAuAKpuugGJIXnrRajGZ5ygACC0AcwCujTMETNMkZMHI10AKADGjbNlAMmLLAB5KAGg0A+UAG8FoS6CUFscoUvJ5MACgtWPNBABMCSAJ6gcvzYxM5wAFygLtB43uhRBqBI6BhIUdBCcABGWACUoAC8JnSG7h6WXr7QAdYAFpBI3jTQUTFxCaCUBcVdANoAuuWgBaYAvgoTCjCSxAhK6KAAotDkkOSB5h6Q3hlZuZgA3KDQCHCddlKCx0oNTS1RK2sbg4eTCsq25OyM5MysmCKJ3QAHcNL9-lg3JYXKZQDtDKdzqAxo9Vut-IYUmkeiZYfDvIizugsal0Cgxnl9O4XM4BGjnv5cURMAIAHS3RrNFgKPJvBQbAAOi005AAch0gUL0CQfloAaB0AAPf7edSiyGYHT8THw6DENh0EwAfg0oAyZKOHxAXTq6Gwi18Jzw3xBzgA1oZyHagv5heoaNg8EgMEE8KAAAasDCYCMCv0i34S+IAfSE0HdsRBhEKvuFsuIEO0gOVqvVRYBOnTmbw2cMMANgKNoFNdHNwJjb2l4IoyfQKejbFz0oLFbYpZYap7f2LOkHmHr+sNJrNFs71rAAAksIsQegAOQ0UCId0wASgBCgIRrJAYvOLXBwRoIQFOUAAJXQ5CEmGglAT7DOOw6aOPg0DYNSNqfoKjBKGekY6hGoAgusdQXrE3psN2MAaG22R4OmU4oUgKBKAgQgOtYXCCs+awEIqmCcDgbIKFB6AwfM8ERohyGoehLp2q+AE4ZhVHQN46xgcgvEkceCDuos6ygDQ36-pG84Rix3aijodBthO4nlvKug6oYOpGCYubDHpioqpORkzpWpl6o2xgtquHZYPy2lJh0ABMQI6SOpCFsZmBGEAA)

**Expected Result**
I'd expect some way to ignore cases like this, i.e. `any` as type parameter for another type in an `extends` position.

**Actual Result**
```
Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
```
I've marked the error location in a comment in Repro.

**Additional Info**

None

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.2` |
| `@typescript-eslint/parser`        | `4.29.2` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `12.20.2` |

**Comment by bradzacher** [comment]
This rule operates solely on the AST - so it does not understand what the type specifically represents - just the code you've written in that place.

Additionally we don't want to add special cases to the rule. Special cases are surprising - if you don't know the rule has special cases then it just looks like the rule has missed a case. This makes people distrust linting, which is a net negative for the linting ecosystem.

If you're intentionally using an `any` - instead just use a disable comment:
```ts
type FlatNode =
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- this is safe because the ignored types don't leak into the conditional result
  typeof flattener extends MatTreeFlattener<any, infer F, any>
    ? F
    : never;
```
Disable comments are great because they are clear - they explicitly explain to any readers why you're bypassing the lint rule and using an `any`.
